broadinstitute/single_cell_portal_core
# Requiring raw count matrices in upload wizard, sync UI/UX (SCP-3492, SCP-3585)

This update expands on the backend work for requiring raw count matrices by enforcing the upload/sync of a raw count matrix before allowing a study owner to add a processed expression matrix.  This update also splits out raw & processed matrix files into separate tabs in the upload wizard, and blocks access to the processed tab until a raw counts matrix has been uploaded, or the user requests an exemption.  This update also adds a new field to the expression file info form for mapping the associations between processed matrix files and their "parent" raw count files.

This new UX will display a transparent overlay div with a disclaimer notifying the user that raw counts files are now required.  Successful upload of a raw file will then remove the overlay and allow access to the processed upload forms.  In addition, the radio buttons for "Is this a raw count matrix" are hard-coded in the upload wizard, and cannot be changed (this is not the case on sync forms, which still allow the user to select either value).  Raw count forms will have the `units` select input, and processed forms will instead render the React select component for selecting the corresponding raw count file.  Upon successful upload/sync of any raw count matrix, these select menus will automatically update to include any new raw files.

New overlay blocking processed uploads:
![Screen Shot 2021-08-30 at 11 01 34 AM](https://user-images.githubusercontent.com/729968/131360208-85e010d0-5bef-46bf-88e9-e3439a8ec392.png)

This update also adds raw counts information to Mixpanel reporting at the conclusion of all ingest jobs.

MANUAL TESTING:
1. Pull this branch and run any pending migrations with `./rails_local_setup.rb && source config/secrets/.source_env.bash && rails db:migrate`
2. Boot server as normal, sign in with an admin account, and set the value for the `raw_counts_required_frontend` and `raw_counts_required_backend` feature flags to `true` for that account
3. Create a new study, or use an existing one with no expression matrix files
4. In the upload wizard, confirm that clicking the 2nd tab for "Processed Matrix" displays the disclaimer and does not allow uploading a matrix file
5. Back in the raw matrix tab, upload the matrix from `test/test_files/expression_matrix_example.txt`
6. After the upload succeeds and the confirmation modal loads, confirm you can now upload a processed matrix on the 2nd tab
7. In the processed upload form, confirm that the select menu for `Corresponding raw count file` now contains the entry for `expression_matrix_example.txt`
8. Upload the file from `test/test_files/expression_matrix_example_2.txt` as a processed matrix, setting the association for the raw count matrix
9. In the rails console, load the above study and confirm that the association has been set correctly on the matrix files:
```
study = Study.last # if you used an existing study, load it by id/accession here
raw_matrix = study.expression_matrix_files.first
processed_matrix = study.expression_matrix_files.last
processed_matrix.associated_matrix_files(:raw).first == raw_matrix
=> true
raw_matrix.associated_matrix_files(:processed).first == processed_matrix
=> true
```
10. (optional) Back in the admin console, turn off the raw counts feature flags for your user account
11. (optional) Create a new study, and confirm that the processed matrix tab does not have the overlay/disclaimer present

This PR satisfies SCP-3492 and SCP-3585.
